internal/provisioner/kops_provisioner.go

```diff
@@ -0,0 +1,113 @@
+	vpnCIDRList             []string
+	owner                   string
+	useExistingAWSResources bool
+	resourceUtil            *utils.ResourceUtil
+	logger                  log.FieldLogger
+	store                   model.InstallationDatabaseStoreInterface
+	kopsCache               map[string]*kops.Cmd
+}
+
+// NewKopsProvisioner creates a new KopsProvisioner.
+func NewKopsProvisioner(s3StateStore, owner string, useExistingAWSResources bool, allowCIDRRangeList, vpnCIDRList []string,
+	resourceUtil *utils.ResourceUtil, logger log.FieldLogger, store model.InstallationDatabaseStoreInterface) *KopsProvisioner {
+	logger = logger.WithField("provisioner", "kops")
+
+	return &KopsProvisioner{
+		s3StateStore:            s3StateStore,
+		useExistingAWSResources: useExistingAWSResources,
+		allowCIDRRangeList:      allowCIDRRangeList,
+		vpnCIDRList:             vpnCIDRList,
+		logger:                  logger,
+		resourceUtil:            resourceUtil,
+		owner:                   owner,
+		store:                   store,
+		kopsCache:               make(map[string]*kops.Cmd),
+	}
+}
+
+// Teardown cleans up cached kops provisioner data.
+func (provisioner *KopsProvisioner) Teardown() {
+	provisioner.logger.Debug("Performing kops provisioner cleanup")
+	for name, kops := range provisioner.kopsCache {
+		provisioner.logger.Debugf("Cleaning up kops cache for %s", name)
+		kops.Close()
+	}
+}
+
+// getKopsClusterConfigLocationFromCache returns the cached kubecfg for a k8s
+// cluster. If the config is not cached, it is fetched with kops.
+func (provisioner *KopsProvisioner) getCachedKopsClusterKubecfg(name string, logger log.FieldLogger) (string, error) {
+	kopsClient, err := provisioner.getCachedKopsClient(name, logger)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get cached kops client")
+	}
+
+	return kopsClient.GetKubeConfigPath(), nil
+}
+
+func (provisioner *KopsProvisioner) getCachedKopsClient(name string, logger log.FieldLogger) (*kops.Cmd, error) {
+	if kopsClient, ok := provisioner.kopsCache[name]; ok {
+		logger.Debugf("Using cached kops client for %s", name)
+		kopsClient.SetLogger(logger)
+		return kopsClient, nil
+	}
+
+	logger.Debugf("Building kops client cache for %s", name)
+	kopsClient, err := kops.New(provisioner.s3StateStore, logger)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create kops wrapper")
+	}
+	err = kopsClient.ExportKubecfg(name)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to export kubecfg")
+	}
+
+	provisioner.kopsCache[name] = kopsClient
+	logger.Debugf("Kops config cached at %s for %s", kopsClient.GetKubeConfigPath(), name)
+
+	return kopsClient, nil
+}
+
+func (provisioner *KopsProvisioner) invalidateCachedKopsClient(name string, logger log.FieldLogger) error {
+	kopsClient, ok := provisioner.kopsCache[name]
+	if !ok {
+		logger.Errorf("Could not find kops client cache for %s to invalidate", name)
+		return errors.Errorf("could not find kops client cache for %s to invalidate", name)
```

Comment:
Is this necessarily an error? I can't think of a case where a caller would want to invalidate the cache and be unable to proceed if the item is missing, as that may indicate that it is already invalidated. Logging the error out makes sense to me but I'm not sure returning an error does.

Reply:
Good question. The reason I did the log and the return of the error is so that the method could be more concretely tested in unit tests. Also, in the future, I could see us maybe needing to know with certainty that the cache was properly invalidated and this sets us up for that.